File: modules/wem-webapp/src/main/webapp/admin/common/js/content/page/region/RegionPlaceableComponentFactory.ts
module api.content.page.region {

    export class RegionPlaceableComponentFactory {

        private static PART_COMPONENT_CLASS_NAME = "PartComponent";

        private static IMAGE_COMPONENT_CLASS_NAME = "ImageComponent";

        private static LAYOUT_COMPONENT_CLASS_NAME = "LayoutComponent";

        public static create(componentAsDataSet: api.data.DataSet): api.content.page.PageComponent<api.content.page.TemplateKey> {

            if (componentAsDataSet.getName() == RegionPlaceableComponentFactory.PART_COMPONENT_CLASS_NAME) {
                return new api.content.page.part.PartComponentBuilder().fromDataSet(componentAsDataSet).build();
            }
            else if (componentAsDataSet.getName() == RegionPlaceableComponentFactory.IMAGE_COMPONENT_CLASS_NAME) {
                return new api.content.page.image.ImageComponentBuilder().fromDataSet(componentAsDataSet).build();
            }
            else if (componentAsDataSet.getName() == RegionPlaceableComponentFactory.LAYOUT_COMPONENT_CLASS_NAME) {
                return new api.content.page.layout.LayoutComponentBuilder().fromDataSet(componentAsDataSet).build();
            }
            else {
                throw new Error("Not a component that can be placed in a Region: " + componentAsDataSet.getName());
            }
        }
    }

}

File: modules/wem-api/src/main/java/com/enonic/wem/api/content/page/region/RegionPlaceableComponentFactory.java
package com.enonic.wem.api.content.page.region;


import com.enonic.wem.api.content.page.PageComponent;
import com.enonic.wem.api.content.page.image.ImageComponent;
import com.enonic.wem.api.content.page.layout.LayoutComponent;
import com.enonic.wem.api.content.page.part.PartComponent;
import com.enonic.wem.api.data.DataSet;

import static com.enonic.wem.api.content.page.image.ImageComponent.newImageComponent;
import static com.enonic.wem.api.content.page.layout.LayoutComponent.newLayoutComponent;
import static com.enonic.wem.api.content.page.part.PartComponent.newPartComponent;

class RegionPlaceableComponentFactory
{
    static PageComponent create( final DataSet dataSet )
    {
        final String classSimpleName = dataSet.getName();
        if ( LayoutComponent.class.getSimpleName().equals( classSimpleName ) )
        {
            return newLayoutComponent().from( dataSet ).build();
        }
        else if ( ImageComponent.class.getSimpleName().equals( classSimpleName ) )
        {
            return newImageComponent().from( dataSet ).build();
        }
        else if ( PartComponent.class.getSimpleName().equals( classSimpleName ) )
        {
            return newPartComponent().from( dataSet ).build();
        }
        else
        {
            throw new IllegalArgumentException( "Not a component that can be placed in a Region: " + classSimpleName );
        }
    }
}
